File: data/fetch.py
"""Pulls the live MP-level MPLADS allocation dataset from the official
pre-login dashboard at https://mplads.mospi.gov.in/digigov/dashboard.html.

The site is a ZK-framework app: its "REST" endpoints are actually
session/desktop-bound AU handlers, not stateless REST, so a plain HTTP
client can't call them directly. We drive a real (headless) browser to
load the dashboard, capture the network responses the page itself makes,
and save the raw JSON payloads. This is the same data a user's browser
receives — nothing here is synthesized.
"""

import json
import re
from datetime import datetime, timezone
from pathlib import Path

from playwright.sync_api import sync_playwright

DASHBOARD_URL = "https://mplads.mospi.gov.in/digigov/dashboard.html"
RAW_DIR = Path(__file__).parent / "raw"
RAW_DIR.mkdir(exist_ok=True)


def fetch() -> None:
    captured: dict[str, str] = {}

    def on_response(response):
        url = response.url
        if "PreLoginDashboardData/getTilesData" in url and response.request.method == "POST":
            captured["tiles"] = response.text()
        elif "PreLoginDashboardData/getTilesReportData" in url and response.request.method == "POST":
            captured["report"] = response.text()

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()
        page.on("response", on_response)
        page.goto(DASHBOARD_URL, wait_until="networkidle", timeout=60000)
        page.wait_for_timeout(2000)
        browser.close()

    if "tiles" not in captured or "report" not in captured:
        raise RuntimeError(
            f"Did not capture expected API responses. Got keys: {list(captured)}"
        )

    tiles = json.loads(captured["tiles"])
    report_wrapper = json.loads(captured["report"])
    # The "Allocated Limit" value is itself a JSON string (escaped) — the
    # ZK grid encodes its data source as a nested JSON string.
    allocated_key = next(k for k in report_wrapper if "Allocated" in k)
    mps = json.loads(report_wrapper[allocated_key])

    fetched_at = datetime.now(timezone.utc).isoformat()

    (RAW_DIR / "tiles.json").write_text(
        json.dumps({"fetched_at": fetched_at, "data": tiles}, indent=2), encoding="utf-8"
    )
    (RAW_DIR / "mp_allocations.json").write_text(
        json.dumps({"fetched_at": fetched_at, "data": mps}, indent=2), encoding="utf-8"
    )

    print(f"Fetched {len(mps)} MP allocation records and national tiles.")
    print(f"Saved to {RAW_DIR}")


if __name__ == "__main__":
    fetch()

File: data/fetch.mjs
// Pulls the live MP-level MPLADS allocation dataset from the official
// pre-login dashboard at https://mplads.mospi.gov.in/digigov/dashboard.html.
//
// The site is a ZK-framework app: its "REST" endpoints are actually
// session/desktop-bound AU handlers, not stateless REST, so a plain HTTP
// client can't call them directly (confirmed: POSTing to them without a
// live ZK desktop session returns a 500). We drive a real headless
// browser to load the dashboard and capture the network responses the
// page itself receives — the same JSON a user's browser gets. Nothing
// here is synthesized.

import { chromium } from "playwright";
import { mkdirSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const RAW_DIR = join(__dirname, "raw");
mkdirSync(RAW_DIR, { recursive: true });

const DASHBOARD_URL = "https://mplads.mospi.gov.in/digigov/dashboard.html";

async function fetchData() {
  const browser = await chromium.launch({ headless: true, channel: "chrome" });
  const page = await browser.newPage();

  const captured = {};
  page.on("response", async (response) => {
    const url = response.url();
    if (response.request().method() !== "POST") return;
    if (url.includes("PreLoginDashboardData/getTilesData")) {
      captured.tiles = await response.text();
    } else if (url.includes("PreLoginDashboardData/getTilesReportData")) {
      captured.report = await response.text();
    }
  });

  await page.goto(DASHBOARD_URL, { waitUntil: "load", timeout: 60000 });

  const deadline = Date.now() + 30000;
  while ((!captured.tiles || !captured.report) && Date.now() < deadline) {
    await page.waitForTimeout(500);
  }
  await browser.close();

  if (!captured.tiles || !captured.report) {
    throw new Error(
      `Did not capture expected API responses. Got keys: ${Object.keys(captured)}`
    );
  }

  const tiles = JSON.parse(captured.tiles);
  const reportWrapper = JSON.parse(captured.report);
  const allocatedKey = Object.keys(reportWrapper).find((k) =>
    k.includes("Allocated")
  );
  const mps = JSON.parse(reportWrapper[allocatedKey]);

  const fetchedAt = new Date().toISOString();

  writeFileSync(
    join(RAW_DIR, "tiles.json"),
    JSON.stringify({ fetched_at: fetchedAt, data: tiles }, null, 2)
  );
  writeFileSync(
    join(RAW_DIR, "mp_allocations.json"),
    JSON.stringify({ fetched_at: fetchedAt, data: mps }, null, 2)
  );

  console.log(`Fetched ${mps.length} MP allocation records and national tiles.`);
  console.log(`Saved to ${RAW_DIR}`);
}

fetchData().catch((err) => {
  console.error(err);
  process.exit(1);
});
